node/core/pvf/src/artifacts.rs

```diff
@@ -0,0 +1,296 @@
+	pub fn serialize(&self) -> Vec<u8> {
+		self.encode()
+	}
+
+	/// Deserialize the given byte buffer to an artifact.
+	pub fn deserialize(mut bytes: &[u8]) -> Result<Self, String> {
+		Artifact::decode(&mut bytes).map_err(|e| format!("{:?}", e))
+	}
+}
+
+/// NOTE if we get to multiple engine implementations the artifact ID should include the engine
+/// type as well.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ArtifactId {
+	code_hash: Hash,
+}
+
+impl ArtifactId {
+	const PREFIX: &'static str = "wasmtime_1_";
+
+	pub fn new(code_hash: Hash) -> Self {
+		Self { code_hash }
+	}
+
+	/// Tries to recover the artifact id from the given file name.
+	pub fn from_file_name(file_name: &str) -> Option<Self> {
+		use std::str::FromStr as _;
+
+		let file_name = file_name.strip_prefix(Self::PREFIX)?;
+		let code_hash = Hash::from_str(file_name).ok()?;
+
+		Some(Self { code_hash })
+	}
+
+	/// Returns the expected path to this artifact given the root of the cache.
+	pub fn path(&self, cache_path: &Path) -> PathBuf {
+		let file_name = format!("{}{}", Self::PREFIX, self.code_hash.to_string());
+		cache_path.join(file_name)
+	}
+}
+
+pub enum ArtifactState {
+	/// The artifact is ready to be used by the executor.
+	///
+	/// That means that the artifact should be accessible through the path obtained by the artifact
+	/// id (unless, it was removed externally).
+	Prepared {
+		/// The time when the artifact was the last time needed.
+		///
+		/// This is updated when we get the heads up for this artifact or when we just discover
+		/// this file.
+		last_time_needed: SystemTime,
+	},
+	/// A task to prepare this artifact is scheduled.
+	Preparing,
+}
+
+/// A container of all known artifact ids.
+pub struct Artifacts {
+	artifacts: HashMap<ArtifactId, ArtifactState>,
+}
+
+impl Artifacts {
+	/// Scan the given cache root for the artifacts.
+	///
+	/// The recognized artifacts will be filled in the table and unrecognized will be removed.
+	pub async fn new(cache_path: &Path) -> Self {
+		// Make sure that the cache path directory and all it's parents are created.
+		let _ = async_std::fs::create_dir_all(cache_path).await;
+
+		let artifacts = match scan_for_known_artifacts(cache_path).await {
+			Ok(a) => a,
+			Err(err) => {
+				tracing::warn!(
+					target: LOG_TARGET,
+					"unable to seed the artifacts in memory cache: {:?}. Starting with a clean one",
+					err,
+				);
+				HashMap::new()
+			}
+		};
+
+		Self { artifacts }
+	}
+
+	#[cfg(test)]
+	pub(crate) fn empty() -> Self {
+		Self {
+			artifacts: HashMap::new(),
+		}
+	}
+
+	/// Returns the state of the given artifact by its ID.
+	pub fn artifact_state_mut(&mut self, artifact_id: &ArtifactId) -> Option<&mut ArtifactState> {
+		self.artifacts.get_mut(artifact_id)
+	}
+
+	/// Inform the table about the artifact with the given ID. The state will be set to "preparing".
+	pub fn insert_preparing(&mut self, artifact_id: ArtifactId) {
+		always!(self
```

Comment:
Is this a type of `debug_assert!`? We should add `qed` proofs for all such assertions.

Reply:
Yes which acts as a log in release.

Hm, I am not sure that the case. And I just run a search across the codebase and the status quo for `debug_assert` is that we don't. In fact, most of them go even without comments.

I can still add the proofs if you insist.

Also this is a somewhat a new situation since this is a fusion of debug_assert and log. Which makes me think that this should not contain proofs because they would look weird in logs.

Reply:
We had a conversation with Gav a couple days ago where he said to use the `qed` proofs for debug asserts, so I am basing my review on that

Reply:
Gotcha, will add proper proofs.